// rqlite-js/lib/api/results/data-results.js
"use strict";var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");Object.defineProperty(exports, "__esModule", { value: true });exports.default = void 0;var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));



var _dataResult = _interopRequireDefault(require("./data-result"));
var _dataResultError = _interopRequireDefault(require("./data-result-error")); /**
 * Class for handling many RQLite data api results
 * @module api/results/data-results
 */ /**
 * A class to manage a list of results from an RQLite query or execute API response
 */class DataResults {













  /**
   * The data result list constructor
   * @param {Object} data The data object sent from a RQLite API response
   * @param {Number} data.time The time the API response took to complete
   * @param {Object[{error:String,values:Object}]} data.results The results array
   */
  constructor(data) {/**
     * The time the results took to return from the
     * RQLite api
     * @type {Number}
     */(0, _defineProperty2.default)(this, "time", 0.0); /**
     * The results which is an empty arry to start
     * @type {DataResult[]}
     */(0, _defineProperty2.default)(this, "results", []);this.setApiData(data);} /**
   * Set the api as results
   * @param {Object} data Api data
   */setApiData(data) {if (typeof data !== 'object') {throw new Error('The data argument is required to be an object');
    }
    if (!data.results) {
      throw new Error('The data object is required to have a results property');
    }
    this.time = parseFloat(data.time || 0.0);
    const _data$results = data.results,results = _data$results === void 0 ? [] : _data$results;
    this.results = results.reduce((acc, result) => {
      // If there is an error property this is an error
      if (typeof result === 'object' && result.error) {
        return acc.concat(new _dataResultError.default(result.error));
      }
      const vals = result.values;
      // We don't have values so this is a single result row
      if (!vals) {
        return acc.concat(new _dataResult.default(result));
      }
      // Map the values to DataResult instances
      const dataResults = vals.map((_v, valuesIndex) => new _dataResult.default(result, valuesIndex));
      return acc.concat(dataResults);
    }, []);
  }

  /**
   * Returns true if an instance of DataResultError exists in the results
   * @returns {Boolean} True if a DataResultError instance exists
   */
  hasError() {
    return !!this.getFirstError();
  }

  /**
   * Get the first error that occured
   * @returns {DataResultError|undefined}
   */
  getFirstError() {
    return this.results.find((v) => v instanceof _dataResultError.default);
  }

  /**
   * Get the time the results took
   * @returns {Number} The time the query took
   */
  getTime() {
    return this.time;
  }

  /**
   * Return one result at a specific index or undefined it it does not exist
   * @returns {DataResult|DataResultError|undefined}
   */
  get(index) {
    return this.results[index];
  }

  /**
   * Return the results array
   * @returns {Array<DataResult|DataResultError>}
   */
  getResults() {
    return this.results;
  }

  /**
   * Get the result data list as array of plain objects
   * @returns {Object[]} The data as an array or objects
   */
  toArray() {
    return this.results.map((result) => result.toObject());
  }

  /**
   * Convert the result data list to a JSON string that is an
   * array of objects
   * @returns {String} A JSON string
   */
  toString() {
    const list = this.results.map((result) => result.toObject());
    return JSON.stringify(list);
  }
}exports.default = DataResults;module.exports = exports.default;
