// rqlite-js/lib/api/client/index.js
"use strict";var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");Object.defineProperty(exports, "__esModule", { value: true });exports.createQuery = createQuery;exports.default = void 0;var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));



var _httpRequest = _interopRequireDefault(require("../../http-request"));
var _httpMethods = require("../../http-request/http-methods");function ownKeys(e, r) {var t = Object.keys(e);if (Object.getOwnPropertySymbols) {var o = Object.getOwnPropertySymbols(e);r && (o = o.filter(function (r) {return Object.getOwnPropertyDescriptor(e, r).enumerable;})), t.push.apply(t, o);}return t;}function _objectSpread(e) {for (var r = 1; r < arguments.length; r++) {var t = null != arguments[r] ? arguments[r] : {};r % 2 ? ownKeys(Object(t), !0).forEach(function (r) {(0, _defineProperty2.default)(e, r, t[r]);}) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) {Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r));});}return e;} /**
 * Base API client for RQLite which abstracts the HTTP calls
 * @module api/client
 */ /**
 * @typedef HttpRequestOptions
 * @type {import('../../http-request').HttpRequestOptions}
 */
/**
 * Create the base HTTP query options from RQLite API options
 * @param {Object} [options={}] The RQLite API options
 * @param {String} [options.level] The consistency level
 * @param {String} [options.pretty] Pretty print the response body
 * @param {String} [options.timings] Provide query timings
 * @param {String} [options.atomic] Treat all commands in the request as a single transaction
 * for RQLite v5 and higher
 * @param {String} [options.transaction] Treat all commands in the request as a single transaction
 * for RQLite v4 and lower
 * @returns {Object} The HTTP query
 */
function createQuery(options = {}) {
  const level = options.level,pretty = options.pretty,timings = options.timings,atomic = options.atomic,transaction = options.transaction;

  // Remove all undefined values
  const query = { level, pretty, timings, atomic, transaction };
  return Object.entries(query).reduce((acc, entry) => {
    const _entry = (0, _slicedToArray2.default)(entry, 2),key = _entry[0],val = _entry[1];
    // Only take defined values
    if (typeof val !== 'undefined') {
      acc[key] = val;
    }
    return acc;
  }, {});
}

/**
 * Base API client for RQLite which abstracts the HTTP calls
 * from the user
 */
class ApiClient extends _httpRequest.default {
  /**
   * Perform a RQLite data API get request
   * @param {String} path The path for the request i.e. /db/query
   * @param {String} sql The SQL query
   * @param {HttpRequestOptions} [options={}] RQLite API options
   */
  get(path, sql, options = {}) {var _superprop_getGet = () => super.get,_this = this;return (0, _asyncToGenerator2.default)(function* () {
      const useLeader = options.useLeader;
      if (!path) {
        throw new Error('The path argument is required');
      }
      return _superprop_getGet().call(_this, {
        useLeader,
        uri: path,
        httpMethod: _httpMethods.HTTP_METHOD_GET,
        query: _objectSpread(_objectSpread({}, createQuery(options)), {}, { q: sql })
      });})();
  }

  /**
   * Perform a RQLite data API post request
   * @param {String} path The path for the request i.e. /db/query
   * @param {String} sql The SQL query
   * @param {HttpRequestOptions} [options={}] RQLite API options
   */
  post(path, sql, options = {}) {var _superprop_getPost = () => super.post,_this2 = this;return (0, _asyncToGenerator2.default)(function* () {
      const useLeader = options.useLeader;
      if (!path) {
        throw new Error('The path argument is required');
      }
      return _superprop_getPost().call(_this2, {
        useLeader,
        uri: path,
        httpMethod: _httpMethods.HTTP_METHOD_POST,
        query: createQuery(options),
        body: Array.isArray(sql) ? sql : [sql]
      });})();
  }
}exports.default = ApiClient;
