// rqlite-js/lib/api/backup/index.js
"use strict";var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");Object.defineProperty(exports, "__esModule", { value: true });exports.default = exports.PATH_LOAD = exports.PATH_BACKUP = exports.BACKUP_DATA_FORMAT_SQL = exports.BACKUP_DATA_FORMAT_DUMP = void 0;var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));



var _httpRequest = _interopRequireDefault(require("../../http-request"));
var _contentTypes = require("../../http-request/content-types"); /**
 * Backup api client to perform RQLite back and load operations
 * @module api/backup
 */ /**
 * The RQLite load api path
 */const PATH_LOAD = exports.PATH_LOAD = '/db/load';

/**
 * The RQLite backup api path
 */
const PATH_BACKUP = exports.PATH_BACKUP = '/db/backup';

/**
 * Use plain SQL dump as the back up format
 */
const BACKUP_DATA_FORMAT_SQL = exports.BACKUP_DATA_FORMAT_SQL = 'sql';

/**
 * Use sqlite3 dump as the back up format
 */
const BACKUP_DATA_FORMAT_DUMP = exports.BACKUP_DATA_FORMAT_DUMP = 'dump';

/**
 * Backup api client to perform RQLite back up and load operations
 */
class BackupApiClient extends _httpRequest.default {
  /**
   * Perform a SQL dump backup from the RQLite server
   * @param {String} [format=BACKUP_DATA_FORMAT_DUMP] The backup data format
   * @returns {Stream} The response stream
   */
  backup(format = BACKUP_DATA_FORMAT_DUMP) {var _superprop_getGet = () => super.get,_this = this;return (0, _asyncToGenerator2.default)(function* () {
      const stream = _superprop_getGet().call(_this, {
        headers: {
          // Always sends application/octet-stream from the server in RQLite v4.x
          Accept: _contentTypes.CONTENT_TYPE_APPLICATION_OCTET_STREAM
        },
        query: { fmt: format === BACKUP_DATA_FORMAT_SQL ? format : undefined },
        json: false,
        stream: true,
        uri: PATH_BACKUP,
        useLeader: true
      });
      return stream;})();
  }

  /**
   * Perform a SQL restore by sending data the RQLite server
   * @param {Buffer|String} data The data to be loaded
   * @param {String} [format=BACKUP_DATA_FORMAT_SQL] The backup data format
   * @returns {Stream} The response stream
   */
  load(data, format = BACKUP_DATA_FORMAT_SQL) {var _superprop_getPost = () => super.post,_this2 = this;return (0, _asyncToGenerator2.default)(function* () {
      return _superprop_getPost().call(_this2, {
        body: data,
        headers: {
          // eslint-disable-next-line max-len
          'Content-Type': format === BACKUP_DATA_FORMAT_SQL ? _contentTypes.CONTENT_TYPE_TEXT_PLAIN : _contentTypes.CONTENT_TYPE_APPLICATION_OCTET_STREAM
        },
        json: false,
        stream: true,
        uri: PATH_LOAD,
        useLeader: true
      });})();
  }
}exports.default = BackupApiClient;
